### ISSUE TITLE
Should reject use of covariant or contravariant TypeVar used in an invariant context

### ISSUE THREAD

**Issue by erictraut** [microsoft/pyright#6955]
If a class-scoped TypeVar is defined explicitly as covariant or contravariant but is used in a context that requires invariance, it should be flagged as an error. Pyright already does this if it's used in a base class that requires invariance, but it doesn't do it in all places.

```python
from typing import Generic, TypeVar

T_co = TypeVar("T_co", covariant=True)

class A(list[T_co]):  # Correctly generates error
    pass

class B(Generic[T_co]):
    def method(self) -> list[T_co]:  # Should generate error
        ...
```

**Comment by erictraut** [comment]
In general, neither pyright nor mypy enforce variance for TypeVars outside of protocols. Adding this would create a lot of noise, so I think I'll refrain from adding it, even though it's technically an error. The new Python 3.12 syntax for type parameters (PEP 695) will largely eliminate this problem because variance will be calculated automatically.
